presto-main/src/main/java/com/facebook/presto/sql/planner/GroupingOperationRewriter.java

```diff
@@ -0,0 +1,129 @@
+
+    public GroupingOperationRewriter(QuerySpecification queryNode, Analysis analysis, Metadata metadata, Map<QuerySpecification, Symbol> groupIdSymbols)
+    {
+        this.queryNode = requireNonNull(queryNode, "node is null");
+        this.analysis = requireNonNull(analysis, "analysis is null");
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.groupIdSymbols = requireNonNull(groupIdSymbols, "groupIdSymbols is null");
+    }
+
+    @Override
+    public Expression rewriteExpression(Expression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        return treeRewriter.defaultRewrite(node, context);
+    }
+
+    @Override
+    public Expression rewriteGroupingOperation(GroupingOperation node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        Expression rewrittenExpression = rewriteGroupingOperationToExpression(node, queryNode);
+        if (rewrittenExpression instanceof FunctionCall) {
+            FunctionCall rewrittenFunctionCall = (FunctionCall) rewrittenExpression;
+            List<TypeSignatureProvider> functionTypes = Arrays.asList(
+                    new TypeSignatureProvider(BIGINT.getTypeSignature()),
+                    new TypeSignatureProvider(ListLiteralType.LIST_LITERAL.getTypeSignature()),
+                    new TypeSignatureProvider(ListLiteralType.LIST_LITERAL.getTypeSignature())
+            );
+            resolveFunction(rewrittenFunctionCall, functionTypes, metadata.getFunctionRegistry());
+        }
+
+        return rewrittenExpression;
+    }
+
+    private Expression rewriteGroupingOperationToExpression(GroupingOperation expression, QuerySpecification node)
+    {
+        List<Expression> columnReferences = ImmutableList.copyOf(analysis.getColumnReferences());
+        List<Expression> groupingOrdinals = expression.getGroupingColumns().stream()
+                .map(columnReferences::indexOf)
+                .map(columnOrdinal -> new LongLiteral(Integer.toString(columnOrdinal)))
+                .collect(toImmutableList());
+
+        List<List<Expression>> groupingSetOrdinals;
+        ImmutableList.Builder<List<Expression>> groupingSetOrdinalsBuilder = ImmutableList.builder();
+        for (List<Expression> groupingSet : analysis.getGroupingSets(node)) {
+            groupingSetOrdinalsBuilder.add(groupingSet.stream()
+                    .map(columnReferences::indexOf)
+                    .map(columnOrdinal -> new LongLiteral(Integer.toString(columnOrdinal)))
+                    .collect(toImmutableList()));
+        }
+        groupingSetOrdinals = groupingSetOrdinalsBuilder.build();
+
+        checkState(node.getGroupBy().isPresent(), "GroupBy node must be present");
+        List<GroupingElement> groupingElements = node.getGroupBy().get().getGroupingElements().stream()
+                .filter(element -> element instanceof GroupingSets || element instanceof Rollup || element instanceof Cube)
```

Comment:
Ya the `element` variable can be an instance of `SimpleGroupBy`. It is possible to have `grouping()` with a simple `GROUP BY`, in which case the below optimization of re-writing the `grouping()` function to a 0 literal kicks in.

This section of the code is not well written so I'll refactor it.